src/topology/algebra/nonarchimedean/basic.lean

```diff
@@ -0,0 +1,138 @@
+
+-/
+
+/-- An topological additive group is nonarchimedean if every neighborhood of 0
+  contains an open subgroup. -/
+class nonarchimedean_add_group (G : Type*)
+  [add_group G] [topological_space G] extends topological_add_group G : Prop :=
+(is_nonarchimedean : ∀ U ∈ nhds (0 : G), ∃ V : open_add_subgroup G, (V : set G) ⊆ U)
+
+/-- A topological group is nonarchimedean if every neighborhood of 1 contains an open subgroup. -/
+@[to_additive]
+class nonarchimedean_group (G : Type*)
+  [group G] [topological_space G] extends topological_group G : Prop :=
+(is_nonarchimedean : ∀ U ∈ nhds (1 : G), ∃ V : open_subgroup G, (V : set G) ⊆ U)
+
+/-- An topological ring is non-archimedean if its underlying topological additive
+  group is nonarchimedean. -/
+class nonarchimedean_ring (R : Type*)
+  [ring R] [topological_space R] extends topological_ring R : Prop :=
+(is_nonarchimedean : ∀ U ∈ nhds (0 : R), ∃ V : open_add_subgroup R, (V : set R) ⊆ U)
+
+/-- Every nonarchimedean ring is naturally a nonarchimedean additive group. -/
+@[priority 100] -- see Note [lower instance priority]
+instance nonarchimedean_ring.to_nonarchimedean_add_group
+  (R : Type*) [ring R] [topological_space R] [t: nonarchimedean_ring R] :
+nonarchimedean_add_group R := {..t}
+
+namespace nonarchimedean_group
+
+variables {G : Type*} [group G] [topological_space G] [nonarchimedean_group G]
+variables {H : Type*} [group H] [topological_space H] [topological_group H]
+variables {K : Type*} [group K] [topological_space K] [nonarchimedean_group K]
+
+/-- If a topological group embeds into a nonarchimedean group, then it
+  is nonarchimedean. -/
+@[to_additive nonarchimedean_add_group.nonarchimedean_of_emb]
+lemma nonarchimedean_of_emb (f : G →* H) (emb : open_embedding f) : nonarchimedean_group H :=
+{ is_nonarchimedean := λ U hU, have h₁ : (f ⁻¹' U) ∈ nhds (1 : G), from
+    by {apply emb.continuous.tendsto, rwa is_group_hom.map_one f},
+  let ⟨V, hV⟩ := is_nonarchimedean (f ⁻¹' U) h₁ in
+    ⟨{is_open' := emb.is_open_map _ V.is_open, ..subgroup.map f V},
+      set.image_subset_iff.2 hV⟩ }
+
+/-- An open neighborhood of the identity in the cartesian product of two nonarchimedean groups
+  contains the cartesian product of an open neighborhood in each group. -/
+@[to_additive nonarchimedean_add_group.prod_subset]
+lemma prod_subset {U} (hU : U ∈ nhds (1 : G × K)) :
+  ∃ (V : open_subgroup G) (W : open_subgroup K), (V : set G).prod (W : set K) ⊆ U :=
+begin
+  erw [nhds_prod_eq, filter.mem_prod_iff] at hU,
+  rcases hU with ⟨U₁, hU₁, U₂, hU₂, h⟩,
+  cases is_nonarchimedean _ hU₁ with V hV,
+  cases is_nonarchimedean _ hU₂ with W hW,
+  use V, use W,
+  rw set.prod_subset_iff,
+  intros x hX y hY,
+  refine set.subset.trans (set.prod_mono hV hW) h _,
+  exact set.mem_sep hX hY,
+end
+
+/-- An open neighborhood of the identity in the cartesian square of two nonarchimedean groups
+  contains the cartesian product of an open neighborhood in each group. -/
```

Comment:
You refer to "two nonarchimedean groups" here, but there's only `G`.

Reply:
```suggestion
/-- An open neighborhood of the identity in the cartesian square of a nonarchimedean group
  contains the cartesian square of an open neighborhood in the group. -/
```

Reply:
Fixed